fluent/client/ws_client.go

```diff
@@ -0,0 +1,150 @@
+package client
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+
+	// "fmt"
+
+	"github.com/IBM/fluent-forward-go/fluent/client/ws"
+	"github.com/IBM/fluent-forward-go/fluent/client/ws/ext"
+	"github.com/gorilla/websocket"
+
+	"github.com/tinylib/msgp/msgp"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+type IAMAuthInfo struct {
+	token string
+	mutex sync.RWMutex
+}
+
+// IAMToken returns the current token value. It is thread safe.
+func (ai *IAMAuthInfo) IAMToken() string {
+	ai.mutex.RLock()
+	defer ai.mutex.RUnlock()
+
+	return ai.token
+}
+
+// SetIAMToken updates the token returned by IAMToken(). It is thread safe.
+func (ai *IAMAuthInfo) SetIAMToken(token string) {
+	ai.mutex.Lock()
+	defer ai.mutex.Unlock()
+
+	ai.token = token
+}
+
+func NewIAMAuthInfo(token string) *IAMAuthInfo {
+	return &IAMAuthInfo{token: token}
+}
+
+// WSSession represents a single websocket connection.
+type WSSession struct {
+	ServerAddress
+	Connection ws.Connection
+}
+
+//counterfeiter:generate . WSConnectionFactory
+type WSConnectionFactory interface {
+	New() (ext.Conn, error)
+}
+
+// DefaultWSConnectionFactory is used by the client if no other
+// ConnectionFactory is provided.
+type DefaultWSConnectionFactory struct {
+	ServerAddress
+	AuthInfo *IAMAuthInfo
+}
+
+func (wcf *DefaultWSConnectionFactory) New() (ext.Conn, error) {
+	var (
+		dialer websocket.Dialer
+		header http.Header
+	)
+
+	if wcf.AuthInfo != nil && len(wcf.AuthInfo.IAMToken()) > 0 {
+		header.Add("Authorization", wcf.AuthInfo.IAMToken())
+	}
+
+	conn, _, err := dialer.Dial(wcf.ServerAddress.String(), header)
+	// TODO: dump response, which is second return value from Dial
+
+	return conn, err
+}
+
+// WSClient manages the lifetime of a single websocket connection.
+type WSClient struct {
+	ConnectionFactory WSConnectionFactory
+	ServerAddress
+	AuthInfo          *IAMAuthInfo
+	ConnectionOptions ws.ConnectionOptions
+	Session           *WSSession
+}
```

Comment:
Any reason WSClient doesn't have a constructor? Otherwise all properties need to be exported which I'm not sure you want ... also you wouldn't have to check if dependencies were nil down below.

Reply:
Mostly because `Client` doesn't have a constructor and I was trying to keep the semantics identical.

You'd still have to check the dependencies so long as the fields are public. There's nothing stopping someone from using the struct anyway.

We can add constructors, but I'd rather do that as part of a PR where we do both clients at the same time.

Reply:
Ok, another PR is fine.